NOnion/Network/TorGuard.fs

```diff
@@ -68,99 +68,147 @@ type TorGuard private (client: TcpClient, sslStream: SslStream) =
+            | Some header ->
+                let circuitId =
+                    header
+                    |> Array.take Constants.CircuitIdLength
+                    |> IntegerSerialization.FromBigEndianByteArrayToUInt16
+
+                // Command is only one byte in size
+                let command =
+                    header
+                    |> Array.skip Constants.CommandOffset
+                    |> Array.exactlyOne
+
+                let! maybeBodyLength =
+                    async {
+                        if Command.IsVariableLength command then
+                            let! maybeLengthBytes =
+                                StreamUtil.ReadFixedSize
+                                    sslStream
+                                    Constants.VariableLengthBodyPrefixLength
+
+                            match maybeLengthBytes with
+                            | Some lengthBytes ->
+                                return
+                                    lengthBytes
+                                    |> IntegerSerialization.FromBigEndianByteArrayToUInt16
+                                    |> int
+                                    |> Some
+                            | None -> return None
+                        else
+                            return Constants.FixedPayloadLength |> Some
+                    }
+
+                match maybeBodyLength with
+                | Some bodyLength ->
+                    let! maybeBody =
+                        StreamUtil.ReadFixedSize sslStream bodyLength
+
+                    match maybeBody with
+                    | Some body -> return Some (circuitId, command, body)
+                    | None -> return None
+                | None -> return None
+            | None -> return None
         }
 
     member private self.ReceiveExpected<'T when 'T :> ICell> () : Async<'T> =
         async {
             let expectedCommandType = Command.GetCommandByCellType<'T> ()
 
             //This is only used for handshake process so circuitId doesn't matter
-            let! _circuitId, command, body = self.ReceiveInternal ()
-
-            //FIXME: maybe continue instead of failing?
-            if command <> expectedCommandType then
-                failwith (sprintf "Unexpected msg type %d" command)
-
-            use memStream = new MemoryStream (body)
-            use reader = new BinaryReader (memStream)
-            return Command.DeserializeCell reader expectedCommandType :?> 'T
+            let! maybeMessage = self.ReceiveInternal ()
+
+            match maybeMessage with
+            | None ->
+                return
+                    failwith
+                        "Socket got closed before receiving an expected cell"
+            | Some (_circuitId, command, body) ->
+                //FIXME: maybe continue instead of failing?
+                if command <> expectedCommandType then
+                    failwith (sprintf "Unexpected msg type %d" command)
```

Comment:
nit (to be fixed in extra commit): this is int, right? then let's use %i instead of %d because it's more readable

Reply:
Completely unrelated to PR, but addresed in a new commit